File: src/experiences/session_shells/ermine/login_shell/lib/clock.dart
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math';

import 'package:flutter/material.dart';

import 'time_stringer.dart';

/// System Clock in the Base Shell
class Clock extends StatelessWidget {
  final TimeStringer _time = TimeStringer();

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        return AnimatedBuilder(
          animation: _time,
          builder: (BuildContext context, Widget child) {
            return Container(
              child: Text(
                _time.timeOnly,
                style: TextStyle(
                  color: Colors.white,
                  fontSize: min(
                    constraints.maxWidth / 6.0,
                    constraints.maxHeight / 6.0,
                  ),
                  fontWeight: FontWeight.w200,
                  letterSpacing: 4.0,
                ),
              ),
            );
          },
        );
      },
    );
  }
}

File: src/sys/pkg/bin/pkg-resolver/src/clock.rs
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use fuchsia_zircon as zx;

#[cfg(not(test))]
pub(crate) fn now() -> zx::Time {
    zx::Time::get_monotonic()
}

#[cfg(test)]
pub(crate) use mock::now;

#[cfg(test)]
pub(crate) mod mock {
    use {super::*, std::cell::Cell};

    thread_local!(
        static MOCK_TIME: Cell<zx::Time> = Cell::new(zx::Time::get_monotonic())
    );

    pub fn now() -> zx::Time {
        MOCK_TIME.with(|time| time.get())
    }

    pub fn set(new_time: zx::Time) {
        MOCK_TIME.with(|time| time.set(new_time));
    }
}
